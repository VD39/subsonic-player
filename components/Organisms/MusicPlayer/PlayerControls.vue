<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import PlayPauseButton from '@/components/Molecules/PlayPauseButton.vue';

import RepeatButton from './Controls/RepeatButton.vue';
import ShuffleButton from './Controls/ShuffleButton.vue';

const {
  fastForwardTrack,
  hasNextTrack,
  hasPreviousTrack,
  isPodcast,
  isRadioStation,
  playNextTrack,
  playPreviousTrack,
  rewindTrack,
} = useAudioPlayer();

const rewindFastForwardTitle = {
  fastForward: `Fast forward back ${FAST_FORWARD_TRACK_TIME} seconds`,
  rewind: `Rewind back ${REWIND_TRACK_TIME} seconds`,
};
</script>

<template>
  <div :class="['centerItems', $style.playerControls]">
    <ShuffleButton v-if="!isRadioStation" />

    <ButtonLink
      ref="previousTrack"
      :icon="ICONS.skipBack"
      title="Previous track"
      :disabled="!hasPreviousTrack"
      @click="playPreviousTrack"
    >
      Previous track
    </ButtonLink>

    <ButtonLink
      v-if="isPodcast"
      ref="rewind"
      :icon="ICONS.rewind"
      :title="rewindFastForwardTitle.rewind"
      @click="rewindTrack"
    >
      {{ rewindFastForwardTitle.rewind }}
    </ButtonLink>

    <PlayPauseButton :class="$style.playPauseButton" />

    <ButtonLink
      v-if="isPodcast"
      ref="fastForward"
      :icon="ICONS.fastForward"
      :title="rewindFastForwardTitle.fastForward"
      @click="fastForwardTrack"
    >
      {{ rewindFastForwardTitle.fastForward }}
    </ButtonLink>

    <ButtonLink
      ref="nextTrack"
      :icon="ICONS.skipForward"
      title="Next track"
      :disabled="!hasNextTrack"
      @click="playNextTrack"
    >
      Next track
    </ButtonLink>

    <RepeatButton v-if="!isRadioStation" />
  </div>
</template>

<style module>
.playerControls {
  gap: var(--default-space);
}

.playPauseButton {
  margin: 0 var(--space-4);
  transform: scale(1.5);
}
</style>
