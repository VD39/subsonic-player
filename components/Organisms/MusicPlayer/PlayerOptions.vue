<script setup lang="ts">
import FavouriteButton from '@/components/Molecules/FavouriteButton.vue';

import MediaInformation from './Controls/MediaInformation.vue';
import PlaybackRateButton from './Controls/PlaybackRateButton.vue';
import VolumeControl from './Controls/VolumeControl.vue';

const {
  currentTime,
  currentTrack,
  duration,
  isPodcast,
  isRadioStation,
  isTrack,
} = useAudioPlayer();
</script>

<template>
  <div :class="['centerItems', $style.playerOptions]">
    <p
      v-if="!isRadioStation && duration"
      ref="timeProgress"
      :class="['smallFont', $style.time]"
    >
      {{ secondsToHHMMSS(currentTime) }} / {{ secondsToHHMMSS(duration) }}
    </p>

    <VolumeControl />

    <PlaybackRateButton v-if="isPodcast" />

    <MediaInformation v-if="isTrack || isPodcast" />

    <FavouriteButton
      v-if="'favourite' in currentTrack"
      :id="currentTrack.id"
      :favourite="currentTrack.favourite"
      :type="currentTrack.type"
    />
  </div>
</template>

<style module>
.playerOptions {
  gap: var(--default-space);
  padding-right: var(--default-space);
}

.time {
  flex-shrink: 0;
  margin-right: var(--default-space);
}
</style>
