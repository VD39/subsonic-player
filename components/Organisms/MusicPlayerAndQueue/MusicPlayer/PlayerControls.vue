<script setup lang="ts">
import MainPlayerControls from '@/components/Organisms/MusicPlayerAndQueue/Controls/MainPlayerControls.vue';
import RepeatButton from '@/components/Organisms/MusicPlayerAndQueue/Controls/RepeatButton.vue';
import ShuffleButton from '@/components/Organisms/MusicPlayerAndQueue/Controls/ShuffleButton.vue';

const { isRadioStation } = useAudioPlayer();
</script>

<template>
  <div :class="['centerItems', $style.playerControls]">
    <ShuffleButton v-if="!isRadioStation" />

    <MainPlayerControls />

    <RepeatButton v-if="!isRadioStation" />
  </div>
</template>

<style module>
.playerControls {
  gap: var(--default-space);
}
</style>
