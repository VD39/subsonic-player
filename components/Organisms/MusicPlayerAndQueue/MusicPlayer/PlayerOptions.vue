<script setup lang="ts">
import FavouriteButton from '@/components/Molecules/FavouriteButton.vue';
import MediaInformation from '@/components/Organisms/MusicPlayerAndQueue/Controls/MediaInformation.vue';
import PlaybackRateButton from '@/components/Organisms/MusicPlayerAndQueue/Controls/PlaybackRateButton.vue';
import QueueButton from '@/components/Organisms/MusicPlayerAndQueue/Controls/QueueButton.vue';
import VolumeControl from '@/components/Organisms/MusicPlayerAndQueue/Controls/VolumeControl.vue';

const {
  currentTime,
  currentTrack,
  duration,
  isPodcastEpisode,
  isRadioStation,
} = useAudioPlayer();
</script>

<template>
  <div :class="['centerAll', $style.playerOptions]">
    <p
      v-if="!isRadioStation && duration"
      ref="timeProgress"
      :class="['smallFont', $style.time]"
    >
      {{ secondsToHHMMSS(currentTime) }} / {{ secondsToHHMMSS(duration) }}
    </p>

    <VolumeControl />

    <PlaybackRateButton v-if="isPodcastEpisode" />

    <MediaInformation v-if="!isRadioStation" />

    <FavouriteButton
      v-if="'favourite' in currentTrack"
      :id="currentTrack.id"
      :favourite="currentTrack.favourite"
      :type="currentTrack.type"
    />

    <QueueButton />
  </div>
</template>

<style module>
.playerOptions {
  gap: var(--default-space);
  padding-right: var(--default-space);
}

.time {
  flex-shrink: 0;
  margin-right: var(--default-space);
}
</style>
