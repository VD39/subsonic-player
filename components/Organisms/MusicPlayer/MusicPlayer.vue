<script setup lang="ts">
import ArtistsList from '@/components/Atoms/ArtistsList.vue';
import PlayPauseButton from '@/components/Molecules/PlayPauseButton.vue';
import PreloadImage from '@/components/Molecules/PreloadImage.vue';
import ImageLink from '@/components/Organisms/ImageLink.vue';

import TrackSeeker from './Controls/TrackSeeker.vue';
import PlayerControls from './PlayerControls.vue';
import PlayerOptions from './PlayerOptions.vue';

const { currentTrack, duration, isRadioStation, isTrack } = useAudioPlayer();
</script>

<template>
  <div :class="$style.musicPlayer">
    <TrackSeeker
      v-if="!isRadioStation && duration"
      :class="['hideMobile', $style.trackSeeker]"
    />

    <div :class="['spaceBetween', 'centerItems', $style.inner]">
      <div :class="['centerItems', $style.trackDetails]">
        <ImageLink
          v-if="isTrack && 'albumId' in currentTrack && currentTrack.albumId"
          :to="`/album/${currentTrack.albumId}`"
          :title="currentTrack.name"
          :image="currentTrack.image"
          :class="$style.image"
        />

        <PreloadImage
          v-else
          :image="currentTrack.image"
          :class="$style.image"
        />

        <div>
          <p class="strong clamp mBXS">
            {{ currentTrack.name }}
          </p>

          <ArtistsList
            v-if="'artists' in currentTrack && currentTrack.artists.length"
            :artists="currentTrack.artists"
            class="smallFont clamp"
          />
        </div>
      </div>

      <PlayPauseButton class="hideDesktop" />

      <PlayerControls class="centerAll hideMobile" />

      <PlayerOptions :class="['hideMobile', $style.playerOptions]" />
    </div>
  </div>
</template>

<style module>
.musicPlayer {
  --music-player-bottom: var(--header-height);

  position: fixed;
  inset: auto 0 var(--music-player-bottom);
  z-index: 9;
  height: var(--media-player-height);
  max-height: var(--media-player-height);
  background-color: var(--background-color);
  border-top: 1px solid var(--border-color);
  box-shadow: var(--dark-box-shadow-medium);

  @media (--tablet-up) {
    --music-player-bottom: 0;
  }
}

.trackSeeker {
  position: absolute;
  z-index: 9;
  display: block;
  transform: translateY(-50%);
}

.inner {
  --inner-gap: var(--default-space);

  position: relative;
  gap: var(--inner-gap);
  padding-right: var(--default-space);

  @media (--tablet-up) {
    --inner-gap: calc(var(--default-space) * 2);

    display: grid;
    grid-template-columns: 35% auto 25%;
  }
}

.trackDetails {
  gap: var(--default-space);
}

.image {
  flex-shrink: 0;
  width: var(--media-player-height);
  height: var(--media-player-height);
}

.playerOptions {
  justify-content: end;
}

.playPauseButton,
.nextTrack {
  transform: scale(1);
}
</style>
