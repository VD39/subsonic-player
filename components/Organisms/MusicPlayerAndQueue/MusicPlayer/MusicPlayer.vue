<script setup lang="ts">
import ArtistsList from '@/components/Atoms/ArtistsList.vue';
import MarqueeScroll from '@/components/Atoms/MarqueeScroll.vue';
import PlayPauseButton from '@/components/Molecules/PlayPauseButton.vue';
import PreloadImage from '@/components/Molecules/PreloadImage.vue';
import ImageLink from '@/components/Organisms/ImageLink.vue';
import QueueButton from '@/components/Organisms/MusicPlayerAndQueue/Controls/QueueButton.vue';
import TrackSeeker from '@/components/Organisms/MusicPlayerAndQueue/Controls/TrackSeeker.vue';

import PlayerControls from './PlayerControls.vue';
import PlayerOptions from './PlayerOptions.vue';

const { currentTrack } = useAudioPlayer();
</script>

<template>
  <div :class="$style.musicPlayer">
    <TrackSeeker :class="['desktopOnly', $style.trackSeeker]" />

    <div :class="['spaceBetween', 'centerItems', $style.inner]">
      <div :class="['centerItems', $style.trackDetailsWrapper]">
        <QueueButton :class="['mobileOnly', $style.queueControl]" />

        <ImageLink
          v-if="'albumId' in currentTrack && currentTrack.albumId"
          ref="albumImageLink"
          :class="$style.image"
          :image="currentTrack.image"
          :title="`Go to album ${currentTrack.name}`"
          :to="`/album/${currentTrack.albumId}`"
        />

        <ImageLink
          v-else-if="'podcastId' in currentTrack && currentTrack.podcastId"
          ref="podcastImageLink"
          :class="$style.image"
          :image="currentTrack.image"
          :title="`Go to podcast ${currentTrack.podcastName}`"
          :to="`/podcast/all/${currentTrack.podcastId}`"
        />

        <PreloadImage
          v-else
          :class="$style.image"
          :image="currentTrack.image"
        />

        <div :class="$style.trackDetails">
          <MarqueeScroll>
            <p class="strong mBXS">
              {{ currentTrack.name }}
            </p>
          </MarqueeScroll>

          <MarqueeScroll
            v-if="'artists' in currentTrack && currentTrack.artists.length"
            ref="artistsMarqueeScroll"
          >
            <ArtistsList :artists="currentTrack.artists" class="smallFont" />
          </MarqueeScroll>

          <MarqueeScroll
            v-if="'author' in currentTrack && currentTrack.author"
            ref="authorMarqueeScroll"
          >
            <p>{{ currentTrack.author }}</p>
          </MarqueeScroll>
        </div>
      </div>

      <PlayPauseButton :class="['mobileOnly', $style.playPauseButton]" />

      <PlayerControls class="centerAll desktopOnly" />

      <PlayerOptions :class="['desktopOnly', $style.playerOptions]" />
    </div>
  </div>
</template>

<style module>
.musicPlayer {
  --music-player-bottom: var(--header-height);

  position: fixed;
  inset: auto 0 var(--music-player-bottom);
  z-index: 9;
  height: var(--media-player-height);
  max-height: var(--media-player-height);
  background-color: var(--background-color);
  border-top: 1px solid var(--border-color);
  box-shadow: var(--dark-box-shadow-medium);

  @media (--tablet-up) {
    --music-player-bottom: 0;
  }
}

.trackSeeker {
  position: absolute;
  z-index: 9;
  display: block;
  transform: translateY(-50%);
}

.inner {
  position: relative;

  @media (--tablet-up) {
    display: grid;
    grid-template-columns: 25% auto 25%;
  }
}

.trackDetailsWrapper {
  position: relative;
  gap: var(--default-space);
}

.trackDetails,
.trackDetailsWrapper {
  flex: 1;
  overflow: hidden;
}

.queueControl {
  position: absolute;
  inset: 0;
  z-index: 1;
  opacity: 0;
}

.image,
.playPauseButton {
  flex-shrink: 0;
  width: var(--media-player-height);
  height: var(--media-player-height);
}

.playPauseButton {
  justify-content: center;
}

.playerOptions {
  justify-content: end;
}
</style>
