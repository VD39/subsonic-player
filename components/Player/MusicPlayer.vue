<script setup lang="ts">
import TrackDetails from '@/components/Queue/TrackDetail.vue';
import TrackSeeker from './Controls/TrackSeeker.vue';
import PlayerControls from './PlayerControls.vue';
import PlayerOptions from './PlayerOptions.vue';
import PlayPauseButton from './Controls/PlayPauseButton.vue';

const { currentTrack } = useAudioPlayer();
</script>

<template>
  <div :class="$style.musicPlayer">
    <TrackSeeker :class="$style.seeker" />

    <div :class="$style.inner">
      <TrackDetails :track="currentTrack" in-media-player />

      <div :class="$style.mobile">
        <PlayPauseButton />
      </div>

      <PlayerControls :class="$style.playerControls" />

      <PlayerOptions :class="$style.playerOptions" />
    </div>
  </div>
</template>

<style module>
.musicPlayer {
  position: fixed;
  inset: auto 0 var(--header-height);
  z-index: 3;
  height: var(--media-player-height);
  max-height: var(--media-player-height);
  background-color: var(--background-color);
  border-top: 1px solid var(--border-color);
  box-shadow: var(--dark-box-shadow-medium);

  @media screen and (--tablet-up) {
    bottom: 0;
  }
}

.seeker {
  position: absolute;
  z-index: 3;
  width: 100%;
  transform: translateY(-50%);
}

.inner {
  @mixin align-center;

  gap: var(--media-player-spacing);
  justify-content: space-between;
  padding-right: var(--media-player-spacing);

  @media screen and (--tablet-up) {
    display: grid;
    grid-template-columns: 25% auto 25%;
    gap: calc(var(--media-player-spacing) * 2);
  }
}

.mobile {
  display: flex;

  @media screen and (--tablet-up) {
    display: none;
  }
}

.playerControls {
  justify-content: center;

  @media screen and (--mobile-only) {
    display: none;
  }
}

.playerOptions {
  justify-content: flex-end;

  @media screen and (--mobile-only) {
    display: none;
  }
}
</style>
