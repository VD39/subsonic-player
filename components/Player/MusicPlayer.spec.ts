import type { VueWrapper } from '@vue/test-utils';
import { mount } from '@vue/test-utils';
import { useAudioPlayerMock } from '@/test/useAudioPlayerMock';
import MusicPlayer from './MusicPlayer.vue';

useAudioPlayerMock();

function factory(props = {}) {
  return mount(MusicPlayer, {
    props: {
      ...props,
    },
    global: {
      stubs: {
        teleport: true,
      },
    },
  });
}

describe('MusicPlayer', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
