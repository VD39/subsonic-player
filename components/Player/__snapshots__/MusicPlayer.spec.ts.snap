// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MusicPlayer > matches the snapshot 1`] = `
"<div class="musicPlayer">
  <div class="sliderWrapper seeker seeker" aria-label="Seek" aria-valuetext="00:05 of 00:10">
    <div class="slider">
      <div class="progressWrapper">
        <div class="process" style="width: 0px;"></div>
        <div class="buffer" style="width: 0px;"></div>
      </div>
      <div class="thumb" style="left: -6px;"></div>
      <div class="tooltip" style="left: 0px;">00:05</div>
    </div>
  </div>
  <div class="inner">
    <div class="trackDetails mediaPlayer">
      <div class="imageWrapper">
        <div class="preloadImage">
          <div><span class="placeholder"></span><span><span class="loader"></span><span class="visually-hidden">Loading image</span></span></div><img src="coverArt" alt="Image" style="display: none;">
        </div>
        <!--v-if-->
      </div>
      <div class="trackMeta">
        <p><a>album</a></p>
        <p class="strong">title</p>
        <div class="secondaryMeta">
          <ul class="meta list">
            <li class="listItem" data-test-id="artist-list-item"><a>name</a></li>
            <li class="listItem" data-test-id="artist-list-item"><a>name1</a></li>
          </ul>
          <p class="duration">00:19</p>
        </div>
      </div>
    </div>
    <div class="mobile">
      <div class="playPauseButton"><span><span class="loader"></span><span class="visually-hidden">Loading data</span></span></div>
    </div>
    <div class="playerControls"><button class="iconButton" type="button" title="Switch shuffle off">
        <phshuffle class="icon" aria-hidden="true" size="24" weight="duotone" color="currentColor"></phshuffle><span class="text visually-hidden">Switch shuffle off</span>
      </button><button class="iconButton disabled" disabled="" type="button" title="Previous track">
        <phskipback class="icon" aria-hidden="true" size="24" weight="duotone"></phskipback><span class="text visually-hidden"> Previous track </span>
      </button>
      <!--v-if-->
      <div class="playPauseButton"><span><span class="loader"></span><span class="visually-hidden">Loading data</span></span></div>
      <!--v-if--><button class="iconButton disabled" disabled="" type="button" title="Next track">
        <phskipforward class="icon" aria-hidden="true" size="24" weight="duotone"></phskipforward><span class="text visually-hidden"> Next track </span>
      </button><button class="iconButton" type="button" title="Turn on repeat all.">
        <phrepeat class="icon" aria-hidden="true" size="24" weight="duotone" color="currentColor"></phrepeat><span class="text visually-hidden">Turn on repeat all.</span>
      </button>
    </div>
    <div class="playerOptions">
      <p class="time">00:05 / 00:10</p>
      <div class="volumeControl"><button class="iconButton" type="button" title="Mute">
          <phspeakerhigh class="icon" aria-hidden="true" size="24" weight="duotone"></phspeakerhigh><span class="text visually-hidden">Mute</span>
        </button>
        <div>
          <div class="sliderWrapper volumeInput volumeInput" aria-label="Volume" aria-valuetext="1 of 1">
            <div class="slider">
              <div class="progressWrapper">
                <div class="process" style="width: 0px;"></div>
                <!--v-if-->
              </div>
              <!--v-if-->
              <!--v-if-->
            </div>
          </div>
          <p class="visually-hidden"> Volume at 100% </p>
        </div>
      </div>
      <!--v-if-->
      <!--v-if-->
      <!--v-if--><button class="iconButton" type="button" title="Queue">
        <phqueue class="icon" aria-hidden="true" size="24" weight="duotone"></phqueue><span class="text visually-hidden"> Queue </span>
      </button>
      <teleport-stub to="#teleports">
        <transition-stub name="slide-up-down" appear="false" persisted="false" css="true">
          <!--v-if-->
        </transition-stub>
      </teleport-stub>
    </div>
  </div>
</div>"
`;
